### PR TITLE
Resolves #1476: Ordering refactorings

### LINKED ISSUE #1476: improve modeling of ordering
This should comprise the following bullets:
- separation of requested (required) orderings as opposed to provided ordering properties (of a plan) which is currently one structure which is inadequate
- removal of plan references out of ordering properties. An ordering should be independent of sub plans. This includes _subproperties_ such as _strictly sorted_ or _reverse_ order
- correctly infer all possible orderings when combining orders for complex operators such as intersections and unions

Example of something that does not work correctly:
- context is a distinct union with two legs
- left leg: ordering provided as `a, b, x` and `c=5`
- right leg: ordering provided as `a, c, x` and `b=6`
- currently the code computes a common ordering which is insufficient to implement the distinct union operator. In fact, currently the ordering that is inferred is dependent on the order of the legs. Common ordering inferred for `left UNION right` is `a, c` and common ordering inferred for `right UNION left` is `a, b`. Since `UNION` is a commutative, that is a problem
- the common ordering provided should be `a, b, c, x` and `a, c, b, x`. Both of these orders should be found irregardless of union operand order

### PR DESCRIPTION
**Pull request by normen662**
- removed plan references from `OrderingInfo`
- as `OrderingInfo` is now just an empty shell delegating everything to `Ordering`, `OrderingInfo` has been removed and has been replaced by `Ordering`
- `Ordering` is only used to describe ordering properties going forward
- `RequestedOrdering` is used to describe an interesting ordering or a required ordering but not an order property
- `Ordering` contains additional information about bindings and more
- an `Ordering` either _satisfies_ a `RequestedOrdering` or it does not
- all requirement push rules use `RequestedOrdering`
- the `OrderingProperty` computes an `Ordering`
- unification of computation of reasoning of ordering in matching and in regular planning code
- representation of ordering while reasoning about ordering as `PartialOrder<K>` which is mathematically more sound IMO.

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Ordering.java [hunk 1]
```diff
@@ -133,7 +142,7 @@ public boolean equals(final Object o) {
         if (!(o instanceof Ordering)) {
             return false;
         }
-        final Ordering ordering = (Ordering)o;
+        final var ordering = (Ordering)o;
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Ordering.java:145]: Done. Thanks!

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/RequestedOrdering.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof RequestedOrdering)) {
+            return false;
+        }
+        final RequestedOrdering ordering = (RequestedOrdering)o;
+        return getOrderingKeyParts().equals(ordering.getOrderingKeyParts());
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/RequestedOrdering.java:94]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/MatchInfo.java [hunk 3]
```diff
@@ -136,17 +136,71 @@ private void collectPredicateMappings(@Nonnull PredicateMap.Builder targetBuilde
...
+     *         {@code Optional.empty()} otherwise. The caller should deal with that result accordingly.
+     *         Note that if there are no bound key parts at all, this method will return {@code Optional.of(false)}.
+     */
+    @Nonnull
+    public Optional<Boolean> deriveReverseScanOrder() {
+        var allForward = true;
+        var allReverse = true;
+
+        for (var boundKeyPart : boundKeyParts) {
+            if (boundKeyPart.isReverse() && allForward) {
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/MatchInfo.java:181]: You are right. Changed.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/combinatorics/TopologicalSort.java [hunk 4]
```diff
@@ -0,0 +1,749 @@
...
+ * ({@code (b, a, c, d}). In this case we would like to instruct the iterator to skip all such orderings and continue
+ * iteration at {@code (b, c, a, d)}. Similarly, we want to skip all orderings starting with {@code (b, ...)} once we
+ * encountered the first such ordering. This the iterators created by the provided {@link EnumeratingIterable}
+ * of type {@link EnumeratingIterator} provide a method {@link EnumeratingIterator#skip}
+ * to allow skipping to a given prefix.
+ *
+ */
+@API(API.Status.EXPERIMENTAL)
+@SuppressWarnings({"java:S4738", "java:S3776"})
+public class TopologicalSort {
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/combinatorics/TopologicalSort.java:98]: There are a few reasons:
- this is code that was just moved in this PR so I am not really introducing something new
- I looked at jgrapht (before I added these classes) when I added the matching code and the explain (DOT) code and jgrapht has some disadvantages in general (LPL and some weird design choices) + Guava has a pretty good graph package that we actually went for (in fact I rewrote an already existing JGraphT implementation to use Guava)
- this iterator AFAICT is giving you exactly one topological sort order element-by-element. `TopologicalSort` as implemented in rec layer gives you an iterator over all the topologically valid sort orders (so each item the iterator returns is a list of items in topologically valid order).
- our iterators can skip subspaces of the search space (see `skip(int)`)
- we have two different ways of iteration: one is a backtracking iterator that in general is faster for cases where you don't have that many dependencies (orderings) between items; and Kahn's algorithm which is way better when enumerating topological orders of elements that almost define complete ordering among their items. In almost every prepackaged algorithm that comes with a graph library you'll get Kahn's algorithm which tends to become very slow if your degrees of freedom are high

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/combinatorics/PartialOrder.java [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+import javax.annotation.Nonnull;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class PartialOrder<T> {
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/combinatorics/PartialOrder.java:40]: Done.

**Review by ohadzeliger — APPROVED** [review]

**Review by MMcM — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/combinatorics/EnumeratingIterable.java [hunk 6]
```diff
@@ -70,4 +75,56 @@ public void skip(final int level) {
             return new EmptyIterator();
         }
     }
+
+    /**
+     * An implementation of {@link EnumeratingIterable} that is optimized to work for single item
+     * input sets. Iterators created by this class, however, avoid to build complex state objects
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/combinatorics/EnumeratingIterable.java:81]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/combinatorics/PartialOrder.java [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+        return Objects.hash(getSet(), getDependencyMap());
+    }
+
+    @Nonnull
+    public static <T> PartialOrder<T> empty() {
+        return new PartialOrder<>(ImmutableSet.of(), ImmutableSetMultimap.of());
+    }
+
+    /**
+     * A class allowing to compute a subset of elements in a partial order that are said to be <em>eligible</em>eligible.
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/combinatorics/PartialOrder.java:130]: Done.

**Review by MMcM — APPROVED** [review]

**Pull request merged**